File: app/src/main/java/com/nandrez/hueman/ui/UseCase.java
/*
 * Copyright (c) 2017. nandrez labs, All Rights Reserved
 */
package com.nandrez.hueman.ui;

/**
 * Use cases are the entry points to the domain layer.
 *
 * @param <Q> the request type
 * @param <P> the response type
 */
public abstract class UseCase<Q extends UseCase.RequestValues, P extends UseCase.ResponseValue> {
    
    private Q requestValues;
    private UseCaseCallback<P> callback;
    
    public void setRequestValues(Q requestValues) {
        this.requestValues = requestValues;
    }
    
    public Q getRequestValues() {
        return requestValues;
    }
    
    public UseCaseCallback<P> getCallback() {
        return callback;
    }
    
    public void setCallback(UseCaseCallback<P> callback) {
        this.callback = callback;
    }
    
    void run() {
        executeUseCase(requestValues);
    }
    
    protected abstract void executeUseCase(Q requestValues);
    
    /**
     * Data passed to a request.
     */
    public interface RequestValues {
    }
    
    /**
     * Data received from a request.
     */
    public interface ResponseValue {
    }
    
    public interface UseCaseCallback<R> {
        void onSuccess(R response);
        void onError();
    }
    
}

File: core/src/main/java/com/nandrez/core/lib/UseCase.kt
/*
 * Copyright (c) 2017. nandrez labs, All Rights Reserved
 */
package com.nandrez.core.lib

/**
 * This class represents a specific interaction with the application's network- and/or server-layers
 */
abstract class UseCase<in Q : UseCase.Request,  out R : UseCase.Response> {
    
    /**
     * Executes the UseCase with the given request-values.
     */
    internal fun run(request: Q, callback: Callback<R>) = executeUseCase(request, callback)
    
    protected abstract fun executeUseCase(request: Q, callback: Callback<R>)
    
    /**
     * Data passed to a request.
     */
    interface Request
    
    /**
     * Data received from a request.
     */
    interface Response
    
    /**
     * An interface allowing the caller to act upon success or failure of his/her request.
     */
    interface Callback<in R> {
        fun onSuccess(response: R)
        fun onError()
    }
}


